Add support for UGW 8.x (and OpenWRT in general)

arnout opened the pull request:
This PR replaces #497.

It is based on #511, adds a few more cleanup commits, and then the remaining commits from #497.

Commit:
tools: OpenWRT external toolchain cmake files and builder docker

CMake and Docker files for building prplMesh out-of-tree on OpenWRT based platforms

Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>
Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
framework: bpl: Renamed RDKB BPL to UCI BPL

Renamed the RDKB BPL to UCI BPL as it'll be the default option for
OpenWRT based platforms. Added TARGET_PLATFORM_TYPE for specifying
sub-types (e.g. "ugw", "turris-omnia").

Updated the CMake file for UGW 8.x

Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

Commit:
common: scripts: Add UGW support to the prplmesh_utils script

The BusyBox implementation of pgrep is quite limited. Some (older?)
versions do not display the command line arguments of the executable.
By matching the full command line (executable + arguments) we ensure
the correct behavior of pgrep.

Also, changing permissions of the /tmp files is only needed on Linux
where we have a normal user that wants to be able to remove the files
without sudo.

Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

Commit:
controller: Add UGW support

Signed-off-by: Vitaly Bukhovsky <vitaly@devalore.com>

arnout committed:
common: bwl: remove brcm, wav, wav_fapi implementations

brcm, wav and wav_fapi are not used for any platform. brcm was not even
included in CMakeLists.txt any longer.

Remove these three implementations and update CMakeLists.txt
accordingly.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
agent, controller: remove BEEROCKS_PATENTS

BEEROCKS_PATENTS is not used anywhere, so remove it, as well as the
BUILD_PATENTS option.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
cmake: move common checks and defaults to top-level CMakeLists.txt

Some configurations are repeated in all or most component CMakeLists.txt
files. Move these to the top-level file so they have to be done only
once.

- cmake_minimum_required
- FindPkgConfig
- GNUInstallDirs
- Minimum g++ version 4.8
- Default CMAKE_INSTALL_PREFIX
- Default CMAKE_BUILD_TYPE
- isSystemDir workaround for RPATH

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
cmake: define TMP_PATH and INSTALL_PATH in top-level CMakeLists.txt

TMP_PATH and BEEROCKS_TMP_PATH are defined and used in several places
but have to be consistent. Similari for MAPF_ROOT and
BEEROCKS_INSTALL_PATH.

Move the definition of these to the top-level CMakeLists.txt to make
sure they're defined the same everywhere, and give the variables a
consistent name TMP_PATH and INSTALL_PATH.

While we're at it, use TMP_PATH directly in prplmesh_utils.sh.in instead
of defining an extra PLATFORM_DB_PATH_TEMP.

As a side effect, the TMP_PATH for the framework is now set to
/tmp/beerocks instead of /tmp. Thus, we no longer have the log and other
files spread out in different directories.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
agent: platform_manager: use sMacAddr as key for m_mapArpEntries

The conversion to uint64_t is pretty much useless. Directly use sMacAddr
itself.

This will allow us to remove network_utils::mac_to_uint64().

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
common: bcl: remove network_utils::mac_to_uint64

It is no longer used, so remove it.

It was anyway not working on any supported platform, because the
implementation was only defined for BEEROCKS_UGW and BEEROCKS_BRCM,
neither of which is ever set.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout commented:
## ***This is not ready to merge yet***

I didn't have time to completely remove BEEROCKS_UGW and a few other things, or to think about a better way to avoid TARGET_PLATFORM_TYPE. I should have created this as a draft but accidentally didn't, and it's not possible to go back to draft :-(

arnout committed:
common: bcl: network_utils: remove useless ANDROID ifdef

This code doesn't build on Android, so having an ifdef for it is
useless.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
common: bcl: network_utils: re-run clang-format

The ifdefs removed in the previous two commits apparently confused
clang-format, since now it is able to properly format arp_send(),
icmp_send() and icmp_checksum().

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

tomereli commented on common/beerocks/scripts/CMakeLists.txt:
```diff
@@ -3,8 +3,11 @@ if (TARGET_PLATFORM STREQUAL "linux")
-    set(PLATFORM_DB_PATH "/opt/prplmesh/share")
+    set(PLATFORM_DB_PATH "/opt/prplMesh/share")
     set(PLATFORM_DB_PATH_TEMP "/tmp/beerocks")
+    if (TARGET_PLATFORM_TYPE STREQUAL "ugw")
```
Not needed anymore

tomereli commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -265,6 +267,6 @@ DATA_IFACE=enx00ee22aa28ef
 
 # Export MultiAP libs folder
-export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:@INSTALL_PATH@/lib
+export LD_LIBRARY_PATH=@INSTALL_PATH@/lib:$LD_LIBRARY_PATH:@EXTRA_LD_PATH@
```
`@EXTRA_LD_PATH@` not needed

tomereli commented on controller/CMakeLists.txt:
```diff
@@ -5,13 +5,17 @@ set(BEEROCKS_WLAN1_IFACE "wlan0")
+if (TARGET_PLATFORM STREQUAL "openwrt")
+    # UGW
+    if (TARGET_PLATFORM_TYPE STREQUAL "ugw")
+        add_definitions(-DBEEROCKS_UGW)
```
shouldn't be needed anymore

tomereli commented on controller/CMakeLists.txt:
```diff
@@ -5,13 +5,17 @@ set(BEEROCKS_WLAN1_IFACE "wlan0")
+        add_definitions(-DBEEROCKS_UGW)
+    # Turris Omnia
+    elseif (TARGET_PLATFORM_TYPE STREQUAL "turris-omnia")
+        set(BEEROCKS_WLAN2_IFACE "wlan1")
```
Did we come to a conclusion about this? I suggest we move the interface names to the platform DB, but that of course can be done in a follow-up PR.

tomereli commented:
> ## _**This is not ready to merge yet**_
> I didn't have time to completely remove BEEROCKS_UGW and a few other things, or to think about a better way to avoid TARGET_PLATFORM_TYPE. I should have created this as a draft but accidentally didn't, and it's not possible to go back to draft :-(

Regarding `TARGET_PLATFORM_TYPE` - I don't think we need it at all **eventually** - `TARGET_PLATFORM` (we should probably change the name since it is more of a `TARGET_OS`) can be `openwrt`, `rdkb` or `linux`, but as of now I also don't see a way to remove it completely.
The usage of `TARGET_PLATFORM_TYPE` currently is for:

- scripts - add `/opt/intel/lib` to the linker search path - **not needed anymore**
- controller
  - Turris Omnia - set `BEEROCKS_WLAN2_IFACE` to be wlan1. Should be in platform DB so won't be needed **but in a different PR**
  - `-DBEEROCKS_UGW`
    - used for versioning. Some of it may be generic but **for a different PR**
    - UGW and RDKB use different safec libraries, so it is used in BWL (DWPAL) and BPL (UCI) to differentiate between them. We should require safec library used by UGW always and update our code 
to use safec anyway, and make a generic mechanism for including the correct headers.
    - agent - different configuration options for the conf files. Should be moved to platform DB
  - `-DBEEROCKS_RDKB` - similar to BEEROCKS_UGW


Bottom line, I think we should keep `TARGET_PLATFORM_TYPE` for now, but remove any dependency which is not mandatory. We should definitely create a task to move all non conventional platform related stuff to the platform DB.

arnout committed:
common: bcl: beerocks_version: remove some version details

When printing or logging the version, a lot of details about the
platform are included. This makes sense if prplMesh is distributed as a
standalone tool, to simplify support (the log file will contain details
about the platform, it doesn't have to be asked separately).

However, these details are already pretty platform-specific (many of
them only work on UGW). Also, we intend to have prplMesh integrated in a
platform, not distributed separately, so presumably the platform details
(and more details than what is reported here by prplMesh) will already
be available through other channels.

Therefore, remove all these details from beerocks_version.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
cmake: safec: remove libsafec from include path

The libsafec header files are included as <libsafec/safe_str_lib.h> so
the libsafec part should not be part of the include directory
specification.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
common: bwl: dynamically find safec/slibc library

The dwpal implementation uses safec string functions. These can come
either from the libsafec or the slibc library. Instead of depending on
the platform, dynamically check both and use the correct one.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

Commit:
cmake: add Finddwpal.cmake and use it in bwl

libdwpal.so is not found when compiling out of tree with UGW8.x,
although it exists in STAGING_DIR/usr/lib (standard search path).
The reason for this is because openwrt expects STAGING_DIR environment
variable to be defined before running the build.
To remove the need for defining STAGING_DIR, just add a Finddwpal.cmake
so it will be found explicitely.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

tomereli commented on tools/docker/builder/OpenWRT.Dockerfile:
```diff
@@ -0,0 +1,53 @@
+ARG PRPLMESH_PLATFORM_TOOLCHAIN_PREFIX
+
+# Make sure that all build arguments are supplied
+RUN  test -n "$PRPLMESH_PLATFORM_TYPE" \
```
Need to be removed. I don't have docker in the UGW build server..

tomereli replied:
Will be fixed in a different PR.

tomereli commented on tools/cmake/toolchain/openwrt.cmake:
```diff
@@ -0,0 +1,51 @@
+## General configuration
```
Missing license header

tomereli replied:
Actually it is not set for .cmake files, so skip it.

tomereli commented on tools/cmake/toolchain/openwrt.cmake:
```diff
@@ -0,0 +1,51 @@
+
+## Toolchain Programs
+set(CMAKE_C_COMPILER ${PLATFORM_TOOLCHAIN_DIR}/bin/${PLATFORM_TOOLCHAIN_PREFIX}gcc CACHE STRING "C Compiler")
+set(CMAKE_CXX_COMPILER ${PLATFORM_TOOLCHAIN_DIR}/bin/${PLATFORM_TOOLCHAIN_PREFIX}g++ CACHE STRING "C++ Compiler")
```
Missing line break

Commit:
cmake: move platform specific flags to toplevel

Move all platform specific flags which are defined all over
the place to the top level CMakeLists.txt.
As a side effect, it also unifies and fixes wlan2 agent not
starting in UGW8.x.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
bpl: cmake: align platform db to TMP path changes

Align platform_db definitions to the actual installation folders using
top level CMAKE variables INSTALL_PATH and TMP_PATH.
This solves the issue that on Linux, test_flows.sh fails due to the
repeater falsely identifies itself as a Gateway since it doesn't find
the platform DB file generated by the prplmesh_utils.sh script.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

tomereli reviewed (approved):
Fixed a bunch of issues UGW issues and an issue causing test_flows.sh to fail due to wrong platform db installation path linux after TMP_PATH changes.
Now I don't see a reason not to merge this so we have an official tag to use for the UGW prplmesh feed Makefile which I also prepared and tested, so approving for merge.
- other UGW issues
